sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -414,6 +414,9 @@ object SimplifyConditionals extends Rule[LogicalPlan] with PredicateHelper {
         // these branches can be pruned away
         val (h, t) = branches.span(_._1 != TrueLiteral)
         CaseWhen( h :+ t.head, None)
+
+      case CaseWhen(Seq((cond, trueValue)), elseValue) =>
+        If(cond, trueValue, elseValue.getOrElse(Literal(null, trueValue.dataType)))
```

Comment:
I don't think `If` is faster than `CaseWhen`, can you explain more about "further optimization"?

Reply:
The generated Java code is slightly simpler, but I agree there should not have any performance gain. Being said that, once `CaseWhen` is converted into `If`, this condition expression will be benefited from the optimization rules in `If` which may not be implemented for `CaseWhen` case.

Reply:
> optimization rules in If which may not be implemented for CaseWhen case.

shall we just implement more optimizer rules for CASE WHEN to cover all the cases?

Reply:
Let's revisit this PR later, and we should always try to add CASE WHEN version for parity. 

Here is the one for case when.
https://github.com/apache/spark/pull/21852